### PR TITLE
ARROW-10151: [Python] Add support for MapArray conversion to Pandas

### PR DESCRIPTION
**Pull request by BryanCutler**
This change adds conversion for a `pyarrow.MapArray` to Pandas as a column of lists of tuples, where each tuple is a key/item pair. Unit tests were added for python to verify conversion for Pandas round-trip, chunked arrays and `MapArray` with NULL map and NULL items.

### CODE REVIEW

File: cpp/src/arrow/python/arrow_to_pandas.cc [hunk 1]
```diff
@@ -791,6 +791,114 @@ Status ConvertListsLike(const PandasOptions& options, const ChunkedArray& data,
...
+    item_arrays.emplace_back(map_arr.items());
+  }
+
+  const auto& map_type = checked_cast<const MapType&>(*data.type());
+  const auto& key_type = map_type.key_type();
+  const auto& item_type = map_type.item_type();
+
+  // TODO: Is this needed for key/item?
+  /*
+  if (value_type->id() == Type::DICTIONARY) {
```
Comment (BryanCutler) [inline cpp/src/arrow/python/arrow_to_pandas.cc:811]: This was taken from the ListConversoin, but not totally sure it applies here to the key/item arrays..

Comment (BryanCutler) [reply cpp/src/arrow/python/arrow_to_pandas.cc]: Yes, same applies. Fixed and added tests.

**Comment by BryanCutler** [comment]
Ping @pitrou @jorisvandenbossche to please take a look, thanks!

**Review by pitrou — CHANGES_REQUESTED** [review]
Thanks for doing this. Here are a couple comments.

**Comment by pitrou** [comment]
Also, you'll want to rebase.

File: cpp/src/arrow/python/arrow_to_pandas.cc [hunk 2]
```diff
@@ -791,6 +791,117 @@ Status ConvertListsLike(const PandasOptions& options, const ChunkedArray& data,
...
+    key_type = dense_type;
+  }
+  if (item_type->id() == Type::DICTIONARY) {
+    auto dense_type = checked_cast<const DictionaryType&>(*item_type).value_type();
+    RETURN_NOT_OK(DecodeDictionaries(options.pool, dense_type, &item_arrays));
+    item_type = dense_type;
+  }
+
+  // TODO(ARROW-489): Currently we don't have a Python reference for single columns.
+  //    Storing a reference to the whole Array would be too expensive.
```
[the remark this answers is not in the document]
Comment (BryanCutler) [reply cpp/src/arrow/python/arrow_to_pandas.cc:823]: JIRA is closed now as Not a Problem, will remove here also

File: cpp/src/arrow/python/arrow_to_pandas.cc [hunk 3]
```diff
@@ -791,6 +791,117 @@ Status ConvertListsLike(const PandasOptions& options, const ChunkedArray& data,
...
+
+  // ARROW-3789(wesm): During refactoring I found that unit tests assumed that
```
[the remark this answers is not in the document]
Comment (BryanCutler) [reply cpp/src/arrow/python/arrow_to_pandas.cc:825]: I saw you updated the related code, will do the same here

File: cpp/src/arrow/python/arrow_to_pandas.cc [hunk 4]
```diff
@@ -791,6 +791,117 @@ Status ConvertListsLike(const PandasOptions& options, const ChunkedArray& data,
...
+    for (int64_t i = 0; i < arr->length(); ++i) {
+      if (has_nulls && arr->IsNull(i)) {
+        Py_INCREF(Py_None);
+        *out_values = Py_None;
+      } else {
+        int64_t entry_offset = arr->value_offset(i);
+        int64_t num_maps = arr->value_offset(i + 1) - entry_offset;
+
+        // Build the new list object for the row of maps
+        list_item.reset(PyList_New(0));
```
[the remark this answers is not in the document]
Comment (BryanCutler) [reply cpp/src/arrow/python/arrow_to_pandas.cc:862]: I tried that at first but had trouble with the ref counts. I didn't use the `PyList_SET_ITEM` macro though, so I'll try again

File: cpp/src/arrow/python/arrow_to_pandas.cc [hunk 5]
```diff
@@ -791,6 +791,117 @@ Status ConvertListsLike(const PandasOptions& options, const ChunkedArray& data,
...
+
+        // Add each key/item pair in the row
+        for (int64_t j = 0; j < num_maps; ++j) {
+          // Get key value, key is non-nullable for a valid row
+          auto ptr_key = reinterpret_cast<const char*>(
+              PyArray_GETPTR1(py_keys, chunk_offset + entry_offset + j));
+          key_value.reset(PyArray_GETITEM(py_keys, ptr_key));
+          RETURN_IF_PYERROR();
+
+          if (item_arrays[c]->IsNull(entry_offset + j)) {
```
[the remark this answers is not in the document]
Comment (BryanCutler) [reply cpp/src/arrow/python/arrow_to_pandas.cc:873]: Good point, let me try that

File: cpp/src/arrow/python/arrow_to_pandas.cc [hunk 6]
```diff
@@ -791,6 +791,117 @@ Status ConvertListsLike(const PandasOptions& options, const ChunkedArray& data,
...
+            // Translate the Null to a None
+            Py_INCREF(Py_None);
+            item_value.reset(Py_None);
+          } else {
+            // Get valid value from item array
+            auto ptr_item = reinterpret_cast<const char*>(
+                PyArray_GETPTR1(py_items, chunk_offset + entry_offset + j));
+            item_value.reset(PyArray_GETITEM(py_items, ptr_item));
+            RETURN_IF_PYERROR();
```
[the remark this answers is not in the document]
Comment (BryanCutler) [reply cpp/src/arrow/python/arrow_to_pandas.cc:882]: Hmm, I'm not an expert on the python/numpy c-apis, so I was following `ConvertStruct` here. I thought that `PyArray_GETITEM` might set an error, but looks like it would return null on failure. Is it possible the reset of the `item_value` PyObject could cause an error?

File: cpp/src/arrow/python/arrow_to_pandas.cc (hunk 5 above)
[the remark this answers is not in the document]
Comment (BryanCutler) [reply cpp/src/arrow/python/arrow_to_pandas.cc:873]: Actually, this doesn't seem to quite work. The backing array is numpy so if the dtype is an object, it will return `None`, but if numeric, it will be `nan`. I think we want the conversion to have `None` for all null values right?

File: cpp/src/arrow/python/arrow_to_pandas.cc [hunk 7]
```diff
@@ -791,6 +791,117 @@ Status ConvertListsLike(const PandasOptions& options, const ChunkedArray& data,
...
+        for (int64_t j = 0; j < num_maps; ++j) {
+          // Get key value, key is non-nullable for a valid row
+          auto ptr_key = reinterpret_cast<const char*>(
+              PyArray_GETPTR1(py_keys, chunk_offset + entry_offset + j));
+          key_value.reset(PyArray_GETITEM(py_keys, ptr_key));
+          RETURN_IF_PYERROR();
+
+          if (item_arrays[c]->IsNull(entry_offset + j)) {
+            // Translate the Null to a None
+            Py_INCREF(Py_None);
```
[the remark this answers is not in the document]
Comment (BryanCutler) [reply cpp/src/arrow/python/arrow_to_pandas.cc:875]: Isn't the line below adding another reference to that object?

File: cpp/src/arrow/python/arrow_to_pandas.cc [hunk 8]
```diff
@@ -791,6 +791,117 @@ Status ConvertListsLike(const PandasOptions& options, const ChunkedArray& data,
...
+
+          // Add the key/item pair to the list for the row
+          PyList_Append(list_item.obj(),
+                        PyTuple_Pack(2, key_value.obj(), item_value.obj()));
+          RETURN_IF_PYERROR();
+        }
+
+        *out_values = list_item.obj();
+        // Grant ownership to the resulting array
+        Py_INCREF(*out_values);
```
[the remark this answers is not in the document]
Comment (BryanCutler) [reply cpp/src/arrow/python/arrow_to_pandas.cc:893]: Again, I was following `ConvertStruct()` but I can change it if you prefer

**Comment by BryanCutler** [comment]
Thanks for reviewing @pitrou , I updated but still had a couple questions and I'm not totally sure I got the reference counting right. I'll have to take a closer look tomorrow.

File: cpp/src/arrow/python/arrow_to_pandas.cc (hunk 5 above)
[the remark this answers is not in the document]
Comment (pitrou) [reply cpp/src/arrow/python/arrow_to_pandas.cc:867]: Ah, ok, thank you.

File: cpp/src/arrow/python/arrow_to_pandas.cc (hunk 7 above)
[the remark this answers is not in the document]
Comment (pitrou) [reply cpp/src/arrow/python/arrow_to_pandas.cc:869]: Hmm, right, it should be ok.

File: cpp/src/arrow/python/arrow_to_pandas.cc (hunk 6 above)
[the remark this answers is not in the document]
Comment (pitrou) [reply cpp/src/arrow/python/arrow_to_pandas.cc:876]: Ah, I though `PyArray_GETITEM` was a simple macro but it seems more involved. So it can return an error indeed.

File: cpp/src/arrow/python/arrow_to_pandas.cc [hunk 9]
```diff
@@ -791,6 +791,111 @@ Status ConvertListsLike(PandasOptions options, const ChunkedArray& data,
...
+            RETURN_IF_PYERROR();
+          }
+
+          // Add the key/item pair to the list for the row
+          PyList_SET_ITEM(list_item.obj(), j,
+                          PyTuple_Pack(2, key_value.obj(), item_value.obj()));
+          RETURN_IF_PYERROR();
+        }
+
+        *out_values = list_item.obj();
```
Comment (pitrou) [inline cpp/src/arrow/python/arrow_to_pandas.cc:885]: `*out_values = list_item.detach()`, simply.

**Review by pitrou — APPROVED** [review]

**Comment by pitrou** [comment]
Will merge if CI is ok, thank you.

File: cpp/src/arrow/python/arrow_to_pandas.cc (hunk 9 above)
Comment (BryanCutler) [reply cpp/src/arrow/python/arrow_to_pandas.cc]: Ah yes, thanks for fixing that up

**Comment by BryanCutler** [comment]
Thanks for all the help @pitrou !
